Prepare v3.4.0 (#2585)
realm/realm-js: CHANGELOG.md

x.x.x Release notes (yyyy-MM-dd)
=============================================================
### Enhancements
* Support mirroring of binary files. Thanks to @malice00. ([#2501](https://github.com/realm/realm-js/issues/2501))

### Fixed

> * Fixed an out-of-range bug in Windows builds. ([realm/realm-object-store#832](https://github.com/realm/realm-object-store/issues/832))

Reviewer: Please describe how an end user would experience that.

x.x.x Release notes (yyyy-MM-dd)
=============================================================
### Enhancements
* Support mirroring of binary files. Thanks to @malice00. ([#2501](https://github.com/realm/realm-js/issues/2501))

> * Performance significantly improved when making a query on the property of a linked table, when the property is indexed. ([]())

Reviewer: Missing the PR or issue reference

### Fixed

> * Fixed an out-of-range bug in Windows builds. In debug mode, the exception `can't dereference out of range vector iterator` will be thrown and the process will terminate. ([realm/realm-object-store#832](https://github.com/realm/realm-object-store/issues/832))

Reviewer 2: Shouldn't this be in past tense since the title is, e.g. `will be thrown` should be `would be thrown`?

Reviewer: As the root cause is less interesting, how about:

In debug mode, the exception `can't dereference out of range vector iterator` could be thrown and the process would terminate. ([realm/realm-object-store#832](https://github.com/realm/realm-object-store/issues/832))

Reviewer: How could people get "in debug mode"?

> * Fixed an out-of-range bug in Windows builds. In debug mode, the exception `can't dereference out of range vector iterator` will be thrown and the process will terminate. ([realm/realm-object-store#832](https://github.com/realm/realm-object-store/issues/832))

Reviewer 3: ```suggestion
* Fixed an out-of-range bug in Windows builds. In debug mode, the exception `can't dereference out of range vector iterator` would be thrown and the process would terminate. ([realm/realm-object-store#832](https://github.com/realm/realm-object-store/issues/832))
```

> * A race between extending the file and activity in the encryption layer could lead to crash and corruption. This race has been fixed. The bug may hit on Android, if encryption is in use. It could also affect Linux on file systems where POSIX `prealloc()` is unsupported. ([realm/realm-core#3427](https://github.com/realm/realm-core/pulls/3427))

Reviewer 3: ```suggestion
* A race between extending the file and activity in the encryption layer could lead to crash and corruption. This race has been fixed. The bug could hit on Android, if encryption was in use. It could also affect Linux on file systems where POSIX `prealloc()` was unsupported. ([realm/realm-core#3427](https://github.com/realm/realm-core/pulls/3427))
```

> * A race between extending the file and activity in the encryption layer could lead to crash and corruption. This race has been fixed. The bug may hit on Android, if encryption is in use. It could also affect Linux on file systems where POSIX `prealloc()` is unsupported. ([realm/realm-core#3427](https://github.com/realm/realm-core/pulls/3427))

Reviewer: This has never been reported on RN Android right? Perhaps people just don't use encryption?
Getting back to the template for describing bugs: Why it happened is less interesting than how it can happen. So in this case I would suggest something along these lines:

* On Android, the realm file could become corrupt when using encrypted realms. This has never been reported in this project. ([realm/realm-core#3427](https://github.com/realm/realm-core/pulls/3427))

### Compatibility
* Realm Object Server: 3.23.1 or later.
* APIs are backwards compatible with all previous release of realm in the 2.x.y series.
* File format: Generates Realms with format v9 (Reads and upgrades all previous formats)

### Internal

> * Updated from Realm Core v5.23.5 to v5.23.6.

Reviewer 2: Personally, I'm not a big fan of this format. It is to much clutter compared to the value it provides (IMO), but 🤷‍♂

Author: @cmelchior Please see https://github.com/realm/realm-js/pull/2553#discussion_r332880466
